File: go-actor-chat/internal/room/registry.go
package room

import (
	"sync"

	"github.com/anthdm/hollywood/actor"
)

const actorKind = "room"

// Registry maps roomId → room actor PID, spawning lazily on first use
// (PRD §9). It is the only public API for resolving a room's actor.
type Registry struct {
	mu     sync.Mutex
	engine *actor.Engine
	store  Store
}

func NewRegistry(engine *actor.Engine, store Store) *Registry {
	return &Registry{engine: engine, store: store}
}

// PIDFor returns the PID of the actor for roomID, spawning it if absent.
// Actors evicted after idle self-poison are respawned transparently.
func (r *Registry) PIDFor(roomID string) *actor.PID {
	r.mu.Lock()
	defer r.mu.Unlock()
	if pid := r.engine.Registry.GetPID(actorKind, roomID); pid != nil {
		return pid
	}
	return r.engine.Spawn(NewActor(roomID, r.store), actorKind, actor.WithID(roomID))
}

File: rust-actor-chat/src/actors/registry.rs
use std::collections::HashMap;
  use std::sync::{Arc, Mutex};

  use crate::actors::room::RoomHandle;
  use crate::actors::RoomId;

  #[derive(Clone, Default)]
  pub struct RoomRegistry {
      inner: Arc<Mutex<HashMap<RoomId, RoomHandle>>>,
  }

  impl RoomRegistry {
      pub fn new() -> Self {
          Self::default()
      }

      /// Handle for a room, spawning on first use — or respawning if the previous
      /// actor self-evicted after going idle.
      pub fn handle_for(&self, room_id: &RoomId) -> RoomHandle {
          let mut map = self.inner.lock().expect("registry mutex poisoned");
          if let Some(existing) = map.get(room_id)
              && !existing.is_closed()
          {
              return existing.clone();
          }
          let handle = RoomHandle::spawn(room_id.clone());
          map.insert(room_id.clone(), handle.clone());
          handle
      }
  }


